base/errorshow.jl

```diff
@@ -336,6 +349,52 @@ function showerror_nostdio(err, msg::AbstractString)
     ccall(:jl_printf, Cint, (Ptr{Cvoid},Cstring), stderr_stream, "\n")
 end
 
+function diagnose_arg_mismatch(io::IO, iob::IO, sigstr, constraining_typevars, @nospecialize(sigT), @nospecialize(gotT))
+    if get(io, :color, false) && isa(gotT, DataType) && isa(sigT, DataType) && gotT.name === sigT.name
+        if constraining_typevars !== nothing && isempty(constraining_typevars)
+            constraining_typevars = nothing
+        end
+        print_with_compare(iob, sigT, gotT, ComparisonOptions(color = :light_red,
+            constraining_typevars = constraining_typevars))
+    elseif get(io, :color, false)
+        Base.with_output_color(Base.error_color(), iob) do iob
+            print(iob, "::", sigstr...)
+        end
+    else
+        print(iob, "!Matched::", sigstr...)
+    end
+end
+
+function rename_typevars(@nospecialize(u), d::Dict{TypeVar, TypeVar})
+    if !isa(u,UnionAll)
+        return u
+    end
+    body = rename_typevars(u.body, d)
+    if body === u.body
+        body = u
+    else
+        body = UnionAll(u.var, body)
+    end
+    var = u.var::TypeVar
+    nv = d[var]
+    return UnionAll(nv, body{nv})
+end
+
+function typevars(u::UnionAll)
+    result = isa(u.body, UnionAll) ? typevars(u.body) : TypeVar[]
+    push!(result, u.var)
+    result
+end
+
+struct MismatchedTypeVar
```

Comment:
it seems somewhat unlikely that this should be a type

Reply:
It's a wrapper type to re-use the generic printing for signatures, but with different formatting for the mismatch. Do you have an alternate suggestion?

Reply:
Hm, not really. It seemed like it should be a bit vector, or continuing the structural matching, but I'm not sure.